### PR TITLE
Refactor mach0 Rebasing and Stripping to use RzBinVirtualFile

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**DO NOT SQUASH**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

"arm64e" Mach-O binaries, as found on arm macOS and later iOS, are not read directly but through a proxy that strips away some info so we get nice pointers. See also the comment in `mach0_rebase.c`
Previously, this was achieved by permanently swapping out the io plugin's read function pointer, but now that virtual files exist it can be implemented without such hacks.

**Test plan**

I've added dedicated tests for all kinds of reads that can appear in such a binary before the changes.

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
set to draft because dyldcache also needs tests for this

File: librz/bin/format/mach0/mach0_rebase.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ * In particular, we strip away additional info stored inside of pointers in the binary so we get
+ * the raw pointers out for convenient analysis.
+ */
+
+#include "mach0.h"
+
+#define IS_PTR_AUTH(x) ((x & (1ULL << 63)) != 0)
+#define IS_PTR_BIND(x) ((x & (1ULL << 62)) != 0)
+
+static void rebase_buffer(struct MACH0_(obj_t) * obj, ut64 off, ut8 *buf, int count) {
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0_rebase.c:22]: Please check for `obj` and `buf` "nullness"

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good, less "hackish" now.

**Review by wargio — APPROVED** [review]
i do not like the fact that there are struct around, but looks good.

File: librz/bin/format/mach0/mach0.c [hunk 2]
```diff
@@ -2214,6 +2214,25 @@ static bool __isDataSection(RzBinSection *sect) {
 	return false;
 }
 
+RzList *MACH0_(get_virtual_files)(RzBinFile *bf) {
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c:2217]: Dont you need to check here that `bf` is also not NULL?

File: librz/bin/format/mach0/mach0.c [hunk 3]
```diff
@@ -2214,6 +2214,25 @@ static bool __isDataSection(RzBinSection *sect) {
...
+		vf->buf = MACH0_(new_rebasing_and_stripping_buf)(obj);
+		vf->buf_owned = true;
+		vf->name = strdup(MACH0_VFILE_NAME_REBASED_STRIPPED);
+		rz_list_push(ret, vf);
+	}
+	return ret;
+}
+
 RzList *MACH0_(get_maps)(RzBinFile *bf) {
 	struct MACH0_(obj_t) *bin = bf->o->bin_obj;
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c:2239]: Same here

**Review by XVilka — APPROVED** [review]
Looks good.

File: librz/bin/format/mach0/mach0.c (hunk 2 above)
Comment (thestr4ng3r) [reply librz/bin/format/mach0/mach0.c:2217]: Doesn't hurt at least.

**Pull request merged**